In [1]:
import pandas as pd
from sqlalchemy import text, create_engine

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Знакомство-с-данными" data-toc-modified-id="Знакомство-с-данными-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Знакомство с данными</a></span><ul class="toc-item"><li><span><a href="#Подсчёт-строк-в-таблицах:" data-toc-modified-id="Подсчёт-строк-в-таблицах:-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Подсчёт строк в таблицах:</a></span></li></ul></li><li><span><a href="#Анализ-базы-данных" data-toc-modified-id="Анализ-базы-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Анализ базы данных</a></span><ul class="toc-item"><li><span><a href="#Кол-во-книг-после-01.01.2000" data-toc-modified-id="Кол-во-книг-после-01.01.2000-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Кол-во книг после 01.01.2000</a></span></li><li><span><a href="#Кол-во-обзоров-и-средняя-оценка" data-toc-modified-id="Кол-во-обзоров-и-средняя-оценка-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Кол-во обзоров и средняя оценка</a></span></li><li><span><a href="#Поиск-издательства" data-toc-modified-id="Поиск-издательства-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Поиск издательства</a></span></li><li><span><a href="#Поиск-автора" data-toc-modified-id="Поиск-автора-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Поиск автора</a></span></li><li><span><a href="#Оценки" data-toc-modified-id="Оценки-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Оценки</a></span></li><li><span><a href="#Выводы:" data-toc-modified-id="Выводы:-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Выводы:</a></span></li></ul></li></ul></div>

<font color='Black' size=+2><b>Анализ базы данных читательского сервиса</b></font><br>

Наша компания хочет занять свою долю на рынке читательских сервисов по подписке. Для этого мы только что приобрели известный сервис и сейчас необходимо детально изучить базы данных книг, которые есть в продаже на данный момент.

Это поможет нам лучше спозиционировать себя и предпринять первые шаги по завоеванию внимания читателей.

In [2]:
db_config = {'user': 'praktikum_student', 
'pwd': 'Sdf4$2;d-d30pp',
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, 
'db': 'data-analyst-final-project-db'}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
#для упрощения запуска напишем функцию:
def revelio(query):
    con=engine.connect()
    return pd.io.sql.read_sql(sql=text(query), con = con)

## Знакомство с данными

База данных содержит 5 таблиц. Выведем на экран первые пять строк каждой таблицы:

______________________
Таблица **books** cодержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

In [4]:
query = '''

SELECT *
FROM books
LIMIT 5

'''

revelio(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

________________________
Таблица **authors** cодержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора

In [5]:
query = '''

SELECT *
FROM authors
LIMIT 5

'''

revelio(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

_________________________
Таблица **publishers** cодержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

In [6]:
query = '''

SELECT *
FROM publishers
LIMIT 5

'''

revelio(query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

________________________
Таблица **ratings** cодержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

In [7]:
query = '''

SELECT *
FROM ratings
LIMIT 5

'''

revelio(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

_________________
Таблица **reviews** cодержит данные о пользовательских обзорах на книги:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text— текст обзора.

In [8]:
query = '''

SELECT *
FROM reviews
LIMIT 5

'''

revelio(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### Подсчёт строк в таблицах:

In [9]:
query_1 = '''

SELECT COUNT(book_id) AS books_total
FROM books

'''

query_2 = '''

SELECT COUNT(review_id) AS reviews_total
FROM reviews

'''

query_3 = '''

SELECT COUNT(author_id) AS authors_total
FROM authors

'''

query_4 = '''

SELECT COUNT(rating_id) AS ratings_total
FROM ratings

'''

query_5 = '''

SELECT COUNT(publisher_id) AS publishers_total
FROM publishers

'''

In [10]:
revelio(query_1)

books_total
0         1000

In [11]:
revelio(query_2)

reviews_total
0           2793

In [12]:
revelio(query_3)

authors_total
0            636

In [13]:
revelio(query_4)

ratings_total
0           6456

In [14]:
revelio(query_5)

publishers_total
0               340

**Вывод по разделу:**

Мы изучили имющуюся базу. Вот базовая информация по ней:
- В нашем распоряжении 1000 книг от 636 авторов;
- Книгам выставлено 6456 оценок и суммарно на них написано 2793 отзыва;
- В выпуске этих книг участвовало 340 издательств;

## Анализ базы данных

В этом разделе мы решим следуюшие задачи:

- Посчитаем, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитаем количество обзоров и среднюю оценку;
- Определим издательство, которое выпустило наибольшее число книг толще 50 страниц, тем самым исключив из анализа брошюры;
- Определим автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками;
- Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

### Кол-во книг после 01.01.2000

Посчитайте, сколько книг вышло после 1 января 2000 года;

In [15]:
query = '''

SELECT COUNT(book_id) AS book_cnt
FROM books
WHERE publication_date > '2000-01-01'

'''

revelio(query)

book_cnt
0       819

### Кол-во обзоров и средняя оценка

Выведем список книг с подсчётом кол-ва обзоров на них. Отсортируем таблицу по кол-ву обзоров, а затем по рейтингу:

In [16]:
query = '''

SELECT b.title, 
       COUNT(DISTINCT re.review_id) AS reviews_cnt, 
       ROUND(AVG(r.rating), 2) AS avg_rating
FROM books b

LEFT JOIN reviews re
ON b.book_id = re.book_id

LEFT JOIN  ratings r
ON b.book_id = r.book_id

GROUP BY b.book_id, b.title
ORDER BY reviews_cnt DESC, avg_rating DESC
LIMIT 15

'''

revelio(query)

title  reviews_cnt  avg_rating
0                             Twilight (Twilight  #1)            7        3.66
1   Harry Potter and the Prisoner of Azkaban (Harr...            6        4.41
2   Harry Potter and the Chamber of Secrets (Harry...            6        4.29
3                                      The Book Thief            6        4.26
4                                    The Glass Castle            6        4.21
5                           Outlander (Outlander  #1)            6        4.13
6                 The Hobbit  or There and Back Again            6        4.13
7   The Curious Incident of the Dog in the Night-Time            6        4.08
8   The Lightning Thief (Percy Jackson and the Oly...            6        4.08
9                                 Water for Elephants            6        3.98
10             The Da Vinci Code (Robert Langdon  #2)            6        3.83
11                             The Catcher in the Rye            6        3.83
12                                      The Alchemist            6        3.79
13                                           The Road            6        3.77
14                          The Giver (The Giver  #1)            6        3.75

Ожидаемо, чаще всего отзывы оставляют на мировые бестселлеры - от "Сумерек" до "Алхимика" и "Над пропастью во ржи". 

### Поиск издательства

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц и тем самым исключим из анализа брошюры

In [17]:
query = '''

WITH temp AS (
    SELECT publisher_id,
           book_id
    FROM books
    WHERE num_pages > 50
    )

SELECT p.publisher,
    COUNT(t.book_id)
FROM temp t

JOIN publishers p
ON t.publisher_id = p.publisher_id

GROUP BY p.publisher
ORDER BY COUNT(t.book_id) DESC
LIMIT 3

'''

revelio(query)

publisher  count
0             Penguin Books     42
1                   Vintage     31
2  Grand Central Publishing     25

Наибольшее кол-во книг выпустило издательство `"Penguin Books"` - 42 книги.

### Поиск автора

Определим автора с самой высокой средней оценкой книг. Будем учитывать только книги с 50 и более оценками.

In [18]:
query = '''

WITH temp AS (
    SELECT book_id
    FROM ratings
    GROUP BY book_id
    HAVING COUNT(rating_id) >= 50
    )


SELECT a.author, 
       ROUND(AVG(r.rating), 2) AS avg_rating
FROM books b

JOIN temp t
ON b.book_id = t.book_id

JOIN authors a
ON b.author_id = a.author_id

JOIN ratings r
ON t.book_id = r.book_id

GROUP BY a.author_id, a.author
ORDER BY AVG(r.rating) DESC

LIMIT 1

'''

revelio(query)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

Автором с самым высоким пользовательским рейтингом оказалась Джоан Роулинг. Её работы мы уже видели в ТОП-3 книг по кол-ву обзоров и среднему рейтингу.

### Оценки

- Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [19]:
query = '''

WITH temp AS (
    SELECT username
    FROM ratings
    GROUP BY username
    HAVING COUNT(rating_id) > 48
    )

SELECT COUNT(DISTINCT re.text) / COUNT(DISTINCT re.username) AS avg_review_cnt
FROM ratings r

JOIN reviews re
ON r.username = re.username

RIGHT JOIN temp t ON r.username = t.username


'''

revelio(query)

avg_review_cnt
0              24

Пользователи пишут обзоры не на каждую книгу, которой они ставят оценку. У пользователей с кол-вом оценок более 48 в среднем 24 текстовых отзыва. Возможно, для нас в будущем было бы полезно поощрять столь активных пользователей.

### Выводы:

Мы проанализировали базу данных и получили следующую информацию:

- В нашем распоряжении 1000 книг от 636 авторов;
- Книгам выставлено 6456 оценок и суммарно на них написано 2793 отзыва;
- В выпуске этих книг участвовало 340 издательств;
____________

- 819 книги из нашей базы вышли в свет после 1 янв 2000 года. Значит купленный нашей компанией сервис делал акцент на современной литературе. Классика есть, но её заметно меньше. 
- Издательство "Penguin Books" представлено в нашем ассортименте 42-мя книгами. Это самый большой показатель среди издательств. Возможно, есть смысл расширить сотрудничество и с другими издательствами, т.к. в базе их 340. Это позитивно скажется на ассортименте. 
- Автор с самым высоким пользовательским рейтингом - Джоан Роулинг. Две её книги входят в топ-3 по кол-ву отзывов;
- Чаще всего пользователи оставляют отзывы на мировые бестселлеры - лидером по отзывам являются "Сумерки";
- У пользователей с кол-вом оценок более 48 в среднем 24 текстовых отзыва.